### ISSUE TITLE
Minimax Music 3: Huggingface cache, symlinks, and unsupported tensor source format.

### ISSUE THREAD

**Issue by AlphaMo99** [0xShug0/audio.cpp#274]
If you download a repository with the huggingface cli : 

` hf download audio-cpp/MiniMax-Music3-GGUF`

And use the cache directory AND specifying a file (works if I do not specify the flow_transformer): 

./build/bin/audiocpp_cli   --task gen --session-option minimax_music3.flow_transformer_gguf=**transformer_q8_0.gguf**   --family minimax_music3   --model /usr/share/llama/gguf/models--audio-cpp--MiniMax-Music3-GGUF/snapshots/ed915d0748225e39b2b9b4eab354a20f66e30bc2/   --backend cuda   --text "A bright pop rock song with clean drums and a clear male vocal."   --request-option 'lyrics=[verse] City lights are shining low. I keep moving with the glow. [chorus] Turn it up and let it fly. Sing the melody tonight.'   --request-option duration_sec=20   --request-option num_inference_steps=30   --out output.wav
ggml_cuda_init: found 1 CUDA devices (Total VRAM: 15937 MiB):
  Device 0: Quadro RTX 5000, compute capability 7.5, VMM: yes, VRAM: 15937 MiB
audiocpp_cli failed: **unsupported tensor source format:** /usr/share/llama/gguf/models--audio-cpp--MiniMax-Music3-GGUF/blobs/40ed8a14884e9ac1b40625edd0dc14138c3d3d62803e2b5903f008dec60e61ac

It fails because the GGUF is a symlink to a file without any file extension : 

 ls -l /usr/share/llama/gguf/models--audio-cpp--MiniMax-Music3-GGUF/snapshots/ed915d0748225e39b2b9b4eab354a20f66e30bc2/transformer_q8_0.gguf 
lrwxrwxrwx 1 nobody nobody 76 Aug 18 17:07 /usr/share/llama/gguf/models--audio-cpp--MiniMax-Music3-GGUF/snapshots/ed915d0748225e39b2b9b4eab354a20f66e30bc2/transformer_q8_0.gguf -> ../../blobs/**40ed8a14884e9ac1b40625edd0dc14138c3d3d62803e2b5903f008dec60e61ac**

**Comment by 0xShug0** [comment]
@AlphaMo99 The `preview` branch has been merged into `main`, and I've pushed the fix there. Please try the latest version from `main`.

**Comment by AlphaMo99** [comment]
Ok it works now with  audiocpp_cli, but not yet with audiocpp_server

**Comment by AlphaMo99** [comment]
Moreover the results looks very strange, only 8 secs of music with no distinctive lyrics, and few notes, like oriental style.

`./build/bin/audiocpp_cli   --task gen --session-option minimax_music3.language_model_gguf=language_model_q8_0.gguf   --session-option minimax_music3.flow_transformer_gguf=transformer_q8_0.gguf --session-option minimax_music3.rvq_depth_decoder_gguf=rvq_depth_decoder_q8_0.gguf  --family minimax_music3   --model /usr/share/llama/gguf/models--audio-cpp--MiniMax-Music3-GGUF/snapshots/ed915d0748225e39b2b9b4eab354a20f66e30bc2/   --backend cuda   --text "A bright pop rock song with clean drums and a clear male vocal."   --request-option 'lyrics=[verse] City lights are shining low. I keep moving with the glow. [chorus] Turn it up and let it fly. Sing the melody tonight.'   --request-option duration_sec=20   --request-option num_inference_steps=30   --out output.wav
ggml_cuda_init: found 1 CUDA devices (Total VRAM: 15937 MiB):
  Device 0: Quadro RTX 5000, compute capability 7.5, VMM: yes, VRAM: 15937 MiB
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture
family=minimax_music3
task=gen
mode=offline
audio_out=output.wav`

If I increase the duration to 60s, the same 8 sec file is generated.

**Comment by 0xShug0** [comment]
> Ok it works now with audiocpp_cli, but not yet with audiocpp_server

I tested the server can handle symlink. Please rebuild the server cleanly and try again. If still happens, share the server config you used.

**Comment by 0xShug0** [comment]
> Moreover the results looks very strange, only 8 secs of music with no distinctive lyrics, and few notes, like oriental style.
> 
> `./build/bin/audiocpp_cli --task gen --session-option minimax_music3.language_model_gguf=language_model_q8_0.gguf --session-option minimax_music3.flow_transformer_gguf=transformer_q8_0.gguf --session-option minimax_music3.rvq_depth_decoder_gguf=rvq_depth_decoder_q8_0.gguf --family minimax_music3 --model /usr/share/llama/gguf/models--audio-cpp--MiniMax-Music3-GGUF/snapshots/ed915d0748225e39b2b9b4eab354a20f66e30bc2/ --backend cuda --text "A bright pop rock song with clean drums and a clear male vocal." --request-option 'lyrics=[verse] City lights are shining low. I keep moving with the glow. [chorus] Turn it up and let it fly. Sing the melody tonight.' --request-option duration_sec=20 --request-option num_inference_steps=30 --out output.wav ggml_cuda_init: found 1 CUDA devices (Total VRAM: 15937 MiB): Device 0: Quadro RTX 5000, compute capability 7.5, VMM: yes, VRAM: 15937 MiB ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture ggml_cuda_graph_set_enabled: disabling CUDA graphs due to GPU architecture family=minimax_music3 task=gen mode=offline audio_out=output.wav`
> 
> If I increase the duration to 60s, the same 8 sec file is generated.

Check the info here https://github.com/0xShug0/audio.cpp/issues/263. The duration is the audio frame budget, aligned with the official Python implementation, not the exact duration of the output audio.   Also, for diffusion-style models, you may need to try different seeds and polish prompt to get the best result.

**Comment by AlphaMo99** [comment]
The Gui doesn't detect the model : (just with --ui and filling `Models folder downloads, local detection, and model loading`)

1) If I try to use the Gui to download again the model : error: remote file is not accessible: audio-cpp/audio.cpp-gguf/MiniMax-Music3-GGUF/config.json (404)

2) Or If I create a symlink to the model dir :   
`ln -s /usr/share/llama/gguf/models--audio-cpp--MiniMax-Music3-GGUF/snapshots/ed915d0748225e39b2b9b4eab354a20f66e30bc2/ MiniMax-Music3-GGUF` (I straced the servers it's looking for MiniMax-Music3-GGUF dir, so I created a link, it finds the directory and gguf files in the strace, but something more is missing, it's still not activated in the Gui). And if I try to download it, I got this message
`error: some package files already exist in: /home/xxx/Projet/AI/audio.cpp/build/bin/models/MiniMax-Music3-GGUF (use --overwrite)`

**Comment by AlphaMo99** [comment]
I took the prompt from here : https://huggingface.co/audio-cpp/MiniMax-Music3-GGUF

**Comment by 0xShug0** [comment]
> I took the prompt from here : https://huggingface.co/audio-cpp/MiniMax-Music3-GGUF

Sorry for the confusion. That prompt is only meant as a quick end-to-end check to confirm the model can run. For higher-quality songs, the prompt needs to be more polished. For example, prompts from https://minimax-ai.github.io/music3-demo/ can easily produce much better results.

[official_q8_20s_no_seed.wav](https://github.com/user-attachments/files/31194802/official_q8_20s_no_seed.wav)
[official_q8_20s_seed_0.wav](https://github.com/user-attachments/files/31194804/official_q8_20s_seed_0.wav)
[official_q8_20s_seed_42.wav](https://github.com/user-attachments/files/31194803/official_q8_20s_seed_42.wav)

**Comment by AlphaMo99** [comment]
ok ok, still trying, If I cut and paste the description, lyrics, length of the very first song, (Features/Pop), I have this error : 
`audiocpp_cli failed: MiniMax Music 3 compact semantic sampler has no finite logits`

**Comment by AlphaMo99** [comment]
Maybe a quote " problem, I'm bisecting the pb to find the root cause

**Comment by 0xShug0** [comment]
> ok ok, still trying, If I cut and paste the description, lyrics, length of the very first song, (Features/Pop), I have this error : `audiocpp_cli failed: MiniMax Music 3 compact semantic sampler has no finite logits`

Can you try changing the seed, and also make sure the prompt/lyrics are not being broken by shell quoting?  The demo acutally uses Features/Pop prompt and works fine.  It's hard to pinpoint the root cause without the exact cmd you used, because I can't reproduce the issue. 

https://github.com/user-attachments/assets/45ce838d-6419-40d4-998d-6ac6e59d2dfd

**Comment by AlphaMo99** [comment]
Ok I put the two files in files : desc.txt and lyrics.txt 
`time ./build/bin/audiocpp_cli   --task gen --session-option minimax_music3.language_model_gguf=language_model_q8_0.gguf   --session-option minimax_music3.flow_transformer_gguf=transformer_q8_0.gguf --session-option minimax_music3.rvq_depth_decoder_gguf=rvq_depth_decoder_q8_0.gguf  --family minimax_music3   --model /usr/share/llama/gguf/models--audio-cpp--MiniMax-Music3-GGUF/snapshots/ed915d0748225e39b2b9b4eab354a20f66e30bc2/   --backend cuda   --text "$(cat desc.txt)"   --request-option "lyrics=$(cat lyrics.txt)" --request-option duration_sec=191   --request-option num_inference_steps=30   --out output.wav --seed 442`

It looks like I have the error message ` audiocpp_cli failed: MiniMax Music 3 compact semantic sampler has no finite logits` when the sum size of the two file is more than 1500 characters (+/- not the exact number, I don't know if the limits is in token or characters precisely...)

**Comment by 0xShug0** [comment]
> Ok I put the two files in files : desc.txt and lyrics.txt `time ./build/bin/audiocpp_cli --task gen --session-option minimax_music3.language_model_gguf=language_model_q8_0.gguf --session-option minimax_music3.flow_transformer_gguf=transformer_q8_0.gguf --session-option minimax_music3.rvq_depth_decoder_gguf=rvq_depth_decoder_q8_0.gguf --family minimax_music3 --model /usr/share/llama/gguf/models--audio-cpp--MiniMax-Music3-GGUF/snapshots/ed915d0748225e39b2b9b4eab354a20f66e30bc2/ --backend cuda --text "$(cat desc.txt)" --request-option "lyrics=$(cat lyrics.txt)" --request-option duration_sec=191 --request-option num_inference_steps=30 --out output.wav --seed 442`
> 
> It looks like I have the error message ` audiocpp_cli failed: MiniMax Music 3 compact semantic sampler has no finite logits` when the sum size of the two file is more than 1500 characters (+/- not the exact number, I don't know if the limits is in token or characters precisely...)

Let me check later. I doubt it's the prompt len because in the demo I used **4000+ char caption and 1200 char lyrics.** 

I'm currently end to end testing the fix for the UI issue. Still downloading the GGUFs from the UI...

**Comment by 0xShug0** [comment]
> Ok I put the two files in files : desc.txt and lyrics.txt `time ./build/bin/audiocpp_cli --task gen --session-option minimax_music3.language_model_gguf=language_model_q8_0.gguf --session-option minimax_music3.flow_transformer_gguf=transformer_q8_0.gguf --session-option minimax_music3.rvq_depth_decoder_gguf=rvq_depth_decoder_q8_0.gguf --family minimax_music3 --model /usr/share/llama/gguf/models--audio-cpp--MiniMax-Music3-GGUF/snapshots/ed915d0748225e39b2b9b4eab354a20f66e30bc2/ --backend cuda --text "$(cat desc.txt)" --request-option "lyrics=$(cat lyrics.txt)" --request-option duration_sec=191 --request-option num_inference_steps=30 --out output.wav --seed 442`
> 
> It looks like I have the error message ` audiocpp_cli failed: MiniMax Music 3 compact semantic sampler has no finite logits` when the sum size of the two file is more than 1500 characters (+/- not the exact number, I don't know if the limits is in token or characters precisely...)

Also could you try the default quant combo?

**Comment by 0xShug0** [comment]
@AlphaMo99  UI fix not pushed yet. Could you share desc.txt and lyrics.txt?

**Comment by 0xShug0** [comment]
@AlphaMo99 Check the the attached txt I used and wav. I can use the feature/pop prompt and your cmd to produce the wav.  The original wav is 30mb+ and can't be uploaded to the comment. I compressed it. 

```
./build/debug/bin/audiocpp_cli \
  --task gen \
  --session-option minimax_music3.language_model_gguf=language_model_q8_0.gguf \
  --session-option minimax_music3.flow_transformer_gguf=transformer_q8_0.gguf \
  --session-option minimax_music3.rvq_depth_decoder_gguf=rvq_depth_decoder_q8_0.gguf \
  --family minimax_music3 \
  --model /path/to/MiniMax-Music3-GGUF \
  --backend cuda \
  --text "$(cat /path/to/desc.txt)" \
  --request-option "lyrics=$(cat /path/to/lyrics.txt)" \
  --request-option duration_sec=191 \
  --request-option num_inference_steps=30 \
  --out q8_seed442_duration191.wav \
  --seed 442 \
  --log \
  --log-file q8_seed442_duration191.log \
  --metrics
```

Finshed: 
```
metrics.wall_ms=154424.794601
metrics.audio_duration_ms=171178
metrics.rtf=0.902133
metrics.x_realtime=1.10848
metrics.audio_duration_source=output
metrics.sample_rate=44100
metrics.channels=2
```

[q8_seed442_duration191.mp3](https://github.com/user-attachments/files/31196887/q8_seed442_duration191.mp3)

[desc.txt](https://github.com/user-attachments/files/31196670/desc.txt)
[lyrics.txt](https://github.com/user-attachments/files/31196671/lyrics.txt)

**Comment by AlphaMo99** [comment]
You're right with the Quant Q4 it works, I guess my 16GB vRam is saturated in Q8 if the text is too long ;) the error message is misleading.
I'll double check with nvtop, still generating... takes time.

**Comment by AlphaMo99** [comment]
no way it's not the qwant lol, still checking

**Comment by AlphaMo99** [comment]
ok it is the quant and vRam exhaut that provide ` audiocpp_cli failed: MiniMax Music 3 compact semantic sampler has no finite logits` I was mislead by `-session-option minimax_music3.language_model_gguf=language_model_q8_0.gguf` vs `--session-option minimax_music3.language_model_gguf=language_model_q8_0.gguf`, the first miss a `-` and produce no error message it just ignore the parameter

**Comment by 0xShug0** [comment]
If you want better speed on a 16GB GPU, you probably do not need to use 30 inference steps. Lower values such as 10 or 20 can be much faster, while still producing reasonable results depending on the prompt and seed.

steps=10  wall_ms=96454.1   audio_duration_ms=161530  rtf=0.597129
steps=20  wall_ms=116229    audio_duration_ms=161530  rtf=0.719549
steps=30  wall_ms=133240    audio_duration_ms=161530  rtf=0.824861

[default_step10_seed442.mp3](https://github.com/user-attachments/files/31197461/default_step10_seed442.mp3)
[default_step20_seed442.mp3](https://github.com/user-attachments/files/31197460/default_step20_seed442.mp3)
[default_step30_seed442.mp3](https://github.com/user-attachments/files/31197462/default_step30_seed442.mp3)

**Comment by AlphaMo99** [comment]
ok cli problem is solved (maybe better error message !?), I'll come back when you're ready to test the Ui/Gui side.

**Comment by AlphaMo99** [comment]
Ho, you already pushed the UI change !, 
I confirm it works now if I'm giving the HF cache dir (with a MiniMax-Music3-GGUF -> models--audio-cpp--MiniMax-Music3-GGUF/snapshots/ed915d0748225e39b2b9b4eab354a20f66e30bc2/ symlink )

**Comment by 0xShug0** [comment]
@AlphaMo99 UI fix pushed. 

Tested:

Download from HF via the UI and run: pass

Create a symlink:
`/tmp/audiocpp-symlink-model-root-20260818-1659/MiniMax-Music3-GGUF -> /path/to/MiniMax-Music3-GGUF`
Then set the UI Models folder to:
`/tmp/audiocpp-symlink-model-root-20260818-1659`
pass

When you create symlink for HF cache. The key is the UI Models folder should be a parent directory that contains model package directories, not the model package directory itself.

```
<models-folder>/
  MiniMax-Music3-GGUF/
    config/
    tokenizer/
    *.gguf
```

So you may do 

```
mkdir -p /path/to/audio-models
ln -s /path/to/hf/snapshot /path/to/audio-models/MiniMax-Music3-GGUF
```

Then set the UI Models folder to:

`/path/to/audio-models`
